// angular2-modal/dist/commonjs/framework/ModalInstanceStack.js
"use strict";
var dom_adapter_1 = require('@angular/platform-browser/src/dom/dom_adapter');
/**
 * A dumb stack implementation over an array.
 */
var ModalInstanceStack = (function () {
    function ModalInstanceStack() {
        this._stack = [];
    }
    ModalInstanceStack.prototype.push = function (mInstance) {
        var idx = this._stack.indexOf(mInstance);
        if (idx === -1)
            this._stack.push(mInstance);
        /* TODO: this is wrong for several reasons:
         1) This is a direct DOM access we need to find another way or to separate it.
         2) It not the place for it.
         3) It doesn't care if its a modal inside an element or a wide open one.
         If its inside an element we need to add the 'modal-open' to that element.
         If its wide open we add to the body, we need to traverse the stack every time
         know what's going on and do it.
         */
        var DOM = dom_adapter_1.getDOM();
        if (DOM && this._stack.length === 1) {
            DOM.addClass(DOM.query('body'), 'modal-open');
        }
    };
    /**
     * Push a ModalDialogInstance into the stack and manage it so when it's done
     * it will automatically kick itself out of the stack.
     * @param mInstance
     */
    ModalInstanceStack.prototype.pushManaged = function (mInstance) {
        var _this = this;
        this.push(mInstance);
        mInstance.result
            .then(function () { return _this.remove(mInstance); })
            .catch(function () { return _this.remove(mInstance); });
        // we don't "pop" because we can't know for sure that our instance is the last.
        // In a user event world it will be the last, but since modals can close programmatically
        // we can't tell.
    };
    ModalInstanceStack.prototype.pop = function () {
        this._stack.pop();
    };
    /**
     * Remove a ModalDialogInstance from the stack.
     * @param mInstance
     */
    ModalInstanceStack.prototype.remove = function (mInstance) {
        var idx = this._stack.indexOf(mInstance);
        if (idx > -1)
            this._stack.splice(idx, 1);
        var DOM = dom_adapter_1.getDOM();
        if (DOM && this._stack.length === 0) {
            DOM.removeClass(DOM.query('body'), 'modal-open');
        }
    };
    ModalInstanceStack.prototype.index = function (index) {
        return this._stack[index];
    };
    ModalInstanceStack.prototype.indexOf = function (mInstance) {
        return this._stack.indexOf(mInstance);
    };
    Object.defineProperty(ModalInstanceStack.prototype, "length", {
        get: function () {
            return this._stack.length;
        },
        enumerable: true,
        configurable: true
    });
    return ModalInstanceStack;
}());
exports.ModalInstanceStack = ModalInstanceStack;
