// angular2-modal/dist/commonjs/presets/TwoButtonPreset.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var core_1 = require('@angular/core');
var Utils_1 = require('../framework/Utils');
var MessageModal_1 = require('../modals/MessageModal');
var MessageModalPreset_1 = require('./base/MessageModalPreset');
function createBindings(config) {
    config.buttons = [
        {
            cssClass: config.okBtnClass,
            caption: config.okBtn,
            onClick: function (modalComponent, $event) {
                return modalComponent.dialog.close(true);
            }
        },
        {
            cssClass: config.cancelBtnClass,
            caption: config.cancelBtn,
            onClick: function (modalComponent, $event) {
                return modalComponent.dialog.dismiss();
            }
        }
    ];
    return core_1.ReflectiveInjector.resolve([
        core_1.provide(MessageModal_1.MessageModalContext, { useValue: config })
    ]);
}
/**
 * A Preset for a classic 2 button modal window.
 */
var TwoButtonPreset = (function (_super) {
    __extends(TwoButtonPreset, _super);
    function TwoButtonPreset(modal, defaultValues) {
        if (defaultValues === void 0) { defaultValues = undefined; }
        _super.call(this, Utils_1.extend({
            modal: modal,
            bindings: createBindings,
            okBtn: 'OK',
            okBtnClass: 'btn btn-primary',
            cancelBtn: 'Cancel',
            cancelBtnClass: 'btn btn-default'
        }, defaultValues || {}), [
            'okBtn',
            'okBtnClass',
            'cancelBtn',
            'cancelBtnClass'
        ]);
    }
    return TwoButtonPreset;
}(MessageModalPreset_1.MessageModalPreset));
exports.TwoButtonPreset = TwoButtonPreset;
